tests/infra/runner.py

```diff
@@ -129,17 +129,26 @@ def run(build_directory, get_command, args):
                 tx_rates = infra.rates.TxRates(primary)
                 while True:
                     if not tx_rates.process_next():
+                        stop_waiting = True
                         for i, remote_client in enumerate(clients):
-                            remote_client.wait()
-                            remote_client.print_and_upload_result(args.label, metrics)
-                            remote_client.stop()
-                        break
+                            done = remote_client.check_done()
+                            # all the clients need to be done
+                            completed = "completed" if done else "not completed"
+                            LOG.info(f"Client {i} has {completed} running")
```

Comment:
you can have the if expression directly in the string interpolation and save the slightly awkwardly named intermediary variable.